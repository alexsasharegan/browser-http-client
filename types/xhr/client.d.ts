import { QueryObject } from "../uri/encode";
import { Response } from "./response";
export interface RequestOptions {
    query?: QueryObject;
    data?: RequestData;
    headers?: {
        [key: string]: string;
    };
    responseType?: XMLHttpRequestResponseType;
    timeout?: number;
}
export declare type RequestData = {
    [key: string]: any;
} | FormData | Document;
export declare type RequestBody = string | FormData | Document;
export declare type HeaderObject = {
    [key: string]: string;
};
export interface PromiseResolve<T> {
    (value?: any): any;
    (value?: T): T;
}
export interface PromiseReject {
    (reason?: any): any;
}
export declare class Client {
    static Timeout: number;
    static ResponseType: XMLHttpRequestResponseType;
    static DefaultHeaders: {
        Accept: string;
        "Cache-Control": string;
        "X-Requested-With": string;
    };
    private xhr;
    private method;
    private url;
    private headers;
    private data;
    private query;
    private responseType;
    private timeout;
    private queryAdded;
    constructor(method: string, url: string);
    setHeaders(headers: HeaderObject): void;
    addQueryString(query: string): void;
    addQueryObject(query: QueryObject): void;
    private setData(data);
    private applyHeaders();
    private handleReadyStateChange(resolve, reject);
    private execute(resolve, reject);
    do(): Promise<Response>;
    static make(method: string, url: string, options?: RequestOptions): Client;
    static get(url: string, query?: QueryObject): Promise<Response>;
    static post(url: string, data: RequestData, options?: RequestOptions): Promise<Response>;
    static put(url: string, data: RequestData, options?: RequestOptions): Promise<Response>;
    static patch(url: string, data: RequestData, options?: RequestOptions): Promise<Response>;
    static delete(url: string, query?: QueryObject): Promise<Response>;
    static setResponseType(type: XMLHttpRequestResponseType): void;
    static readonly Async: boolean;
}
