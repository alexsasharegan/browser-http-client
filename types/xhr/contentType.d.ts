export declare const ContentType: {
    json: string;
    text: string;
    html: string;
    form: string;
    multipart: string;
    stream: string;
};
