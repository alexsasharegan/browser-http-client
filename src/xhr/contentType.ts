export const ContentType = {
  json: "application/json",
  text: "text/plain",
  html: "text/html",
  form: "application/x-www-form-urlencoded",
  multipart: "multipart/form-data",
  stream: "application/octet-stream"
}
