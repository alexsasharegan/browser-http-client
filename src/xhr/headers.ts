export type Dictionary<T> = { [key: string]: T };

const ignoreDupeMap: Dictionary<boolean> = {
  age: true,
  authorization: true,
  "content-length": true,
  "content-type": true,
  etag: true,
  expires: true,
  from: true,
  host: true,
  "if-modified-since": true,
  "if-unmodified-since": true,
  "last-modified": true,
  location: true,
  "max-forwards": true,
  "proxy-authorization": true,
  referer: true,
  "retry-after": true,
  "user-agent": true,
};

/**
 * Parse headers into the target object.
 *
 * ```
 * const resHeaders =
 * `Date: Wed, 27 Aug 2014 08:58:49 GMT
 * Content-Type: application/json
 * Connection: keep-alive
 * Transfer-Encoding: chunked`
 *
 * parseHeaders(resHeaders)
 * // Output:
 * { date: 'Wed, 27 Aug 2014 08:58:49 GMT',
 *  'content-type': 'application/json',
 *  connection: 'keep-alive',
 *  'transfer-encoding': 'chunked' }
 * ```
 */
export function parse(
  headers: string,
  target_mut: { [key: string]: string | string[] }
): { [field: string]: string | string[] } {
  if (!headers) {
    return target_mut;
  }

  let line: string;
  let key: string;
  let val: string;
  let i: number;
  let lead_sp = /^\s*/;
  let trail_sp = /\s*$/;

  for (line of headers.split("\n")) {
    i = line.indexOf(":");
    key = line
      .substr(0, i)
      .replace(lead_sp, "")
      .replace(trail_sp, "")
      .toLowerCase();
    val = line
      .substr(i + 1)
      .replace(lead_sp, "")
      .replace(trail_sp, "");

    // Skip empty keys and defined items in the ignore list.
    if (key == "" || (target_mut[key] !== undefined && ignoreDupeMap[key])) {
      continue;
    }

    if (key === "set-cookie") {
      // Normalize to array
      if (!Array.isArray(target_mut[key])) {
        target_mut[key] = [];
      }
      // TS needs a var alloc to infer type, so force the type
      (target_mut[key] as string[]).push(val);
      continue;
    }

    // These keys can have multiple values
    if (target_mut[key] !== undefined) {
      target_mut[key] += ", " + val;
      continue;
    }

    target_mut[key] = val;
  }

  return target_mut;
}

export interface HttpHeaders {
  Accept?: string;
  ["Access-Control-Allow-Origin"]?: string;
  ["Access-Control-Allow-Credentials"]?: string;
  ["Access-Control-Expose-Headers"]?: string;
  ["Access-Control-Max-Age"]?: string;
  ["Access-Control-Allow-Methods"]?: string;
  ["Access-Control-Allow-Headers"]?: string;
  ["Accept-Patch"]?: string;
  ["Accept-Ranges"]?: string;
  Age?: string;
  Allow?: string;
  ["Alt-Svc"]?: string;
  ["Cache-Control"]?: string;
  connection?: string;
  ["Content-Disposition"]?: string;
  ["Content-Encoding"]?: string;
  ["Content-Language"]?: string;
  ["Content-Length"]?: string;
  ["Content-Location"]?: string;
  ["Content-Range"]?: string;
  ["Content-Type"]?: string;
  Date?: string;
  Expires?: string;
  Host?: string;
  ["Last-Modified"]?: string;
  Location?: string;
  Pragma?: string;
  ["Proxy-Authenticate"]?: string;
  ["Public-Key-Pins"]?: string;
  ["Retry-After"]?: string;
  ["Set-Cookie"]?: string;
  ["Strict-Transport-Security"]?: string;
  Trailer?: string;
  ["Transfer-Encoding"]?: string;
  Tk?: string;
  Upgrade?: string;
  Vary?: string;
  Via?: string;
  Warning?: string;
  ["Www-Authenticate"]?: string;
  [header: string]: string | undefined;
}
