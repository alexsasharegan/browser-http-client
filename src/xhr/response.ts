export class Response {
  public data: any
  public status: number
  public statusText:string
  public headers: object
  public xhr: XMLHttpRequest
}
