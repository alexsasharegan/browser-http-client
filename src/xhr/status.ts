export enum Code {
  // HTTP status codes as registered with IANA.
  // See: http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
  StatusContinue = 100, // RFC 7231, 6.2.1
  StatusSwitchingProtocols = 101, // RFC 7231, 6.2.2
  StatusProcessing = 102, // RFC 2518, 10.1

  StatusOK = 200, // RFC 7231, 6.3.1
  StatusCreated = 201, // RFC 7231, 6.3.2
  StatusAccepted = 202, // RFC 7231, 6.3.3
  StatusNonAuthoritativeInfo = 203, // RFC 7231, 6.3.4
  StatusNoContent = 204, // RFC 7231, 6.3.5
  StatusResetContent = 205, // RFC 7231, 6.3.6
  StatusPartialContent = 206, // RFC 7233, 4.1
  StatusMultiStatus = 207, // RFC 4918, 11.1
  StatusAlreadyReported = 208, // RFC 5842, 7.1
  StatusIMUsed = 226, // RFC 3229, 10.4.1

  StatusMultipleChoices = 300, // RFC 7231, 6.4.1
  StatusMovedPermanently = 301, // RFC 7231, 6.4.2
  StatusFound = 302, // RFC 7231, 6.4.3
  StatusSeeOther = 303, // RFC 7231, 6.4.4
  StatusNotModified = 304, // RFC 7232, 4.1
  StatusUseProxy = 305, // RFC 7231, 6.4.5
  _ = 306, // RFC 7231, 6.4.6 (Unused)
  StatusTemporaryRedirect = 307, // RFC 7231, 6.4.7
  StatusPermanentRedirect = 308, // RFC 7538, 3

  StatusBadRequest = 400, // RFC 7231, 6.5.1
  StatusUnauthorized = 401, // RFC 7235, 3.1
  StatusPaymentRequired = 402, // RFC 7231, 6.5.2
  StatusForbidden = 403, // RFC 7231, 6.5.3
  StatusNotFound = 404, // RFC 7231, 6.5.4
  StatusMethodNotAllowed = 405, // RFC 7231, 6.5.5
  StatusNotAcceptable = 406, // RFC 7231, 6.5.6
  StatusProxyAuthRequired = 407, // RFC 7235, 3.2
  StatusRequestTimeout = 408, // RFC 7231, 6.5.7
  StatusConflict = 409, // RFC 7231, 6.5.8
  StatusGone = 410, // RFC 7231, 6.5.9
  StatusLengthRequired = 411, // RFC 7231, 6.5.10
  StatusPreconditionFailed = 412, // RFC 7232, 4.2
  StatusRequestEntityTooLarge = 413, // RFC 7231, 6.5.11
  StatusRequestURITooLong = 414, // RFC 7231, 6.5.12
  StatusUnsupportedMediaType = 415, // RFC 7231, 6.5.13
  StatusRequestedRangeNotSatisfiable = 416, // RFC 7233, 4.4
  StatusExpectationFailed = 417, // RFC 7231, 6.5.14
  StatusTeapot = 418, // RFC 7168, 2.3.3
  StatusUnprocessableEntity = 422, // RFC 4918, 11.2
  StatusLocked = 423, // RFC 4918, 11.3
  StatusFailedDependency = 424, // RFC 4918, 11.4
  StatusUpgradeRequired = 426, // RFC 7231, 6.5.15
  StatusPreconditionRequired = 428, // RFC 6585, 3
  StatusTooManyRequests = 429, // RFC 6585, 4
  StatusRequestHeaderFieldsTooLarge = 431, // RFC 6585, 5
  StatusUnavailableForLegalReasons = 451, // RFC 7725, 3

  StatusInternalServerError = 500, // RFC 7231, 6.6.1
  StatusNotImplemented = 501, // RFC 7231, 6.6.2
  StatusBadGateway = 502, // RFC 7231, 6.6.3
  StatusServiceUnavailable = 503, // RFC 7231, 6.6.4
  StatusGatewayTimeout = 504, // RFC 7231, 6.6.5
  StatusHTTPVersionNotSupported = 505, // RFC 7231, 6.6.6
  StatusVariantAlsoNegotiates = 506, // RFC 2295, 8.1
  StatusInsufficientStorage = 507, // RFC 4918, 11.5
  StatusLoopDetected = 508, // RFC 5842, 7.2
  StatusNotExtended = 510, // RFC 2774, 7
  StatusNetworkAuthenticationRequired = 511 // RFC 6585, 6
}

export const Text = {
  [Code.StatusContinue]: "Continue",
  [Code.StatusSwitchingProtocols]: "Switching Protocols",
  [Code.StatusProcessing]: "Processing",

  [Code.StatusOK]: "OK",
  [Code.StatusCreated]: "Created",
  [Code.StatusAccepted]: "Accepted",
  [Code.StatusNonAuthoritativeInfo]: "Non-Authoritative Information",
  [Code.StatusNoContent]: "No Content",
  [Code.StatusResetContent]: "Reset Content",
  [Code.StatusPartialContent]: "Partial Content",
  [Code.StatusMultiStatus]: "Multi-Status",
  [Code.StatusAlreadyReported]: "Already Reported",
  [Code.StatusIMUsed]: "IM Used",

  [Code.StatusMultipleChoices]: "Multiple Choices",
  [Code.StatusMovedPermanently]: "Moved Permanently",
  [Code.StatusFound]: "Found",
  [Code.StatusSeeOther]: "See Other",
  [Code.StatusNotModified]: "Not Modified",
  [Code.StatusUseProxy]: "Use Proxy",
  [Code.StatusTemporaryRedirect]: "Temporary Redirect",
  [Code.StatusPermanentRedirect]: "Permanent Redirect",

  [Code.StatusBadRequest]: "Bad Request",
  [Code.StatusUnauthorized]: "Unauthorized",
  [Code.StatusPaymentRequired]: "Payment Required",
  [Code.StatusForbidden]: "Forbidden",
  [Code.StatusNotFound]: "Not Found",
  [Code.StatusMethodNotAllowed]: "Method Not Allowed",
  [Code.StatusNotAcceptable]: "Not Acceptable",
  [Code.StatusProxyAuthRequired]: "Proxy Authentication Required",
  [Code.StatusRequestTimeout]: "Request Timeout",
  [Code.StatusConflict]: "Conflict",
  [Code.StatusGone]: "Gone",
  [Code.StatusLengthRequired]: "Length Required",
  [Code.StatusPreconditionFailed]: "Precondition Failed",
  [Code.StatusRequestEntityTooLarge]: "Request Entity Too Large",
  [Code.StatusRequestURITooLong]: "Request URI Too Long",
  [Code.StatusUnsupportedMediaType]: "Unsupported Media Type",
  [Code.StatusRequestedRangeNotSatisfiable]: "Requested Range Not Satisfiable",
  [Code.StatusExpectationFailed]: "Expectation Failed",
  [Code.StatusTeapot]: "I'm a teapot",
  [Code.StatusUnprocessableEntity]: "Unprocessable Entity",
  [Code.StatusLocked]: "Locked",
  [Code.StatusFailedDependency]: "Failed Dependency",
  [Code.StatusUpgradeRequired]: "Upgrade Required",
  [Code.StatusPreconditionRequired]: "Precondition Required",
  [Code.StatusTooManyRequests]: "Too Many Requests",
  [Code.StatusRequestHeaderFieldsTooLarge]: "Request Header Fields Too Large",
  [Code.StatusUnavailableForLegalReasons]: "Unavailable For Legal Reasons",

  [Code.StatusInternalServerError]: "Internal Server Error",
  [Code.StatusNotImplemented]: "Not Implemented",
  [Code.StatusBadGateway]: "Bad Gateway",
  [Code.StatusServiceUnavailable]: "Service Unavailable",
  [Code.StatusGatewayTimeout]: "Gateway Timeout",
  [Code.StatusHTTPVersionNotSupported]: "HTTP Version Not Supported",
  [Code.StatusVariantAlsoNegotiates]: "Variant Also Negotiates",
  [Code.StatusInsufficientStorage]: "Insufficient Storage",
  [Code.StatusLoopDetected]: "Loop Detected",
  [Code.StatusNotExtended]: "Not Extended",
  [Code.StatusNetworkAuthenticationRequired]: "Network Authentication Required"
}
