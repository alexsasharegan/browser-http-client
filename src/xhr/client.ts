import { Ok, Err, Result, expect_never } from "safe-types";
import * as Headers from "./headers";
import * as QStr from "../encoding/qstr";
import { ContentType } from "./contentType";

export interface RequestOptions {
  query?: QStr.QueryObject;
  data?: RequestData;
  headers?: Headers.HttpHeaders;
  responseType?: XMLHttpRequestResponseType;
  timeout?: number;
}

export type RequestData = { [key: string]: any } | FormData | Document;
export type RequestBody = string | FormData | Document;

export interface Response<T = any> {
  data: T;
  status: number;
  statusText: string;
  headers: { [field: string]: string | string[] };
  xhr: XMLHttpRequest;
}

export type Matchable<T> = { match: ClientErrMatcher<T> };

export type ClientErrorStatus<T> = {
  xhr: XMLHttpRequest;
  readonly type: RequestErrType.HttpStatusErr;
  response: Response<T>;
  event: Event;
};

export type ClientErrorAbort = {
  xhr: XMLHttpRequest;
  readonly type: RequestErrType.Abort;
  event: Event;
};

export type ClientErrorProgress = {
  xhr: XMLHttpRequest;
  readonly type: RequestErrType.Timeout;
  event: ProgressEvent;
};

export type ClientErrorError = {
  xhr: XMLHttpRequest;
  readonly type: RequestErrType.XhrErr;
  event: ErrorEvent;
};

export type ClientErrorBase<T> =
  | ClientErrorStatus<T>
  | ClientErrorAbort
  | ClientErrorProgress
  | ClientErrorError;

export type ClientError<T> =
  | Matchable<T> & ClientErrorStatus<T>
  | Matchable<T> & ClientErrorAbort
  | Matchable<T> & ClientErrorProgress
  | Matchable<T> & ClientErrorError;

export type ErrMatchObj<T, U> = {
  [type in RequestErrType]: (err: ClientErrorBase<T>) => U
};

export interface ClientErrMatcher<T> {
  <U>(matcher: ErrMatchObj<T, U>): U;
}

function err_with_matcher<T>(err: ClientErrorBase<T>): ClientError<T> {
  return Object.assign(err, {
    match<U>(matcher: ErrMatchObj<T, U>): U {
      switch (err.type) {
        case RequestErrType.Abort:
          return matcher[RequestErrType.Abort](err);

        case RequestErrType.HttpStatusErr:
          return matcher[RequestErrType.HttpStatusErr](err);

        case RequestErrType.Timeout:
          return matcher[RequestErrType.Timeout](err);

        case RequestErrType.XhrErr:
          return matcher[RequestErrType.XhrErr](err);

        default:
          return expect_never(err, "invalid error type");
      }
    },
  });
}

export enum RequestErrType {
  HttpStatusErr = "HttpStatusErr",
  XhrErr = "XhrErr",
  Timeout = "Timeout",
  Abort = "Abort",
}

function new_response<T>(xhr: XMLHttpRequest): Response<T> {
  let r: Response<T> = Object.create(null);

  r.xhr = xhr;
  r.data = xhr.response;
  r.status = xhr.status;
  r.statusText = xhr.statusText;
  r.headers = Headers.parse(xhr.getAllResponseHeaders(), Object.create(null));

  return r;
}

export interface PromiseResolver<T> {
  (value: T): T;
}

export class Client<T> {
  public static Timeout: number = 0;
  public static ResponseType: XMLHttpRequestResponseType = "json";

  public static DefaultHeaders = {
    Accept: ContentType.Json,
    "X-Requested-With": "XMLHttpRequest",
  };

  private xhr: XMLHttpRequest = new XMLHttpRequest();
  private headers: Headers.HttpHeaders = Object.assign(
    Object.create(null),
    Client.DefaultHeaders
  );
  private data: RequestBody;
  private responseType: XMLHttpRequestResponseType = Client.ResponseType;
  private timeout: number = Client.Timeout;
  private query_added: boolean = false;
  private req_sent: boolean = false;
  private promise: Promise<Result<Response<T>, ClientError<T>>>;
  private xhr_evt_listeners: Array<{
    type: keyof XMLHttpRequestEventMap;
    listener: (
      this: XMLHttpRequest,
      ev: XMLHttpRequestEventMap[keyof XMLHttpRequestEventMap]
    ) => any;
    options?: boolean | AddEventListenerOptions;
  }> = [];
  private upload_evt_listeners: Array<{
    type: keyof XMLHttpRequestEventTargetEventMap;
    listener: (
      this: XMLHttpRequestUpload,
      ev: XMLHttpRequestEventTargetEventMap[keyof XMLHttpRequestEventTargetEventMap]
    ) => any;
    options?: boolean | AddEventListenerOptions;
  }> = [];

  constructor(private method: string, private url: string) {}

  /**
   * Adds an object of headers via `Object.assign`
   * that will be written to the XHR.
   */
  public addHeaders(headers: Headers.HttpHeaders): void {
    Object.assign(this.headers, headers);
  }
  /**
   * Sets a new object of headers that will be written to the XHR.
   */
  public setHeaders(headers: Headers.HttpHeaders): void {
    this.headers = Object.assign(Object.create(null), headers);
  }

  public addQueryStr(query: string): void {
    if (this.query_added) {
      throw new Error("Cannot add query string twice.");
    }
    this.url += "?";
    this.url += query;
    this.query_added = true;
  }

  public addQueryObj(query: QStr.QueryObject): void {
    this.addQueryStr(QStr.encode_query(query));
  }

  private setData(data: RequestData): void {
    if (data instanceof FormData || data instanceof Document) {
      this.data = data;
      return;
    }
    this.data = JSON.stringify(data);
    this.addHeaders({ "Content-Type": ContentType.Json });
  }

  private applyHeaders(): void {
    let header: string;
    for (header in this.headers) {
      if (!this.headers[header]) {
        continue;
      }
      this.xhr.setRequestHeader(header, this.headers[header]);
    }
  }

  /**
   * Returns a function that will cancel the underlying xhr.
   * If the xhr is underway, this will generate an error with type: `Abort`.
   */
  public getCancelToken(): () => void {
    return () => this.xhr.abort();
  }

  /**
   * Register event handlers on the xhr.
   * There's largely no reason to need this since
   * the xhr events are watched internally.
   */
  public addEventListener<K extends keyof XMLHttpRequestEventMap>(
    type: K,
    listener: (this: XMLHttpRequest, ev: XMLHttpRequestEventMap[K]) => any,
    options?: boolean | AddEventListenerOptions
  ): void {
    this.xhr_evt_listeners.push({
      type,
      listener,
      options,
    });
  }

  /**
   * Register event handlers on the xhr upload.
   * Useful for progress events on a file upload.
   */
  public addUploadEventListener<
    K extends keyof XMLHttpRequestEventTargetEventMap
  >(
    type: K,
    listener: (
      this: XMLHttpRequestUpload,
      ev: XMLHttpRequestEventTargetEventMap[K]
    ) => any,
    options?: boolean | AddEventListenerOptions
  ): void {
    this.upload_evt_listeners.push({
      type,
      listener,
      options,
    });
  }

  private handleReadyStateChange(
    resolve: PromiseResolver<Result<Response<T>, ClientError<T>>>,
    event: Event
  ) {
    if (this.xhr.readyState !== XMLHttpRequest.DONE) {
      return;
    }

    let r = new_response<T>(this.xhr);
    if (this.xhr.status < 200 || this.xhr.status >= 400) {
      resolve(
        Err(
          err_with_matcher({
            xhr: this.xhr,
            type: RequestErrType.HttpStatusErr,
            response: r,
            event,
          })
        )
      );
    }

    resolve(Ok(r));
  }

  private execute(
    resolve: PromiseResolver<Result<Response<T>, ClientError<T>>>
  ): void {
    this.xhr.open(this.method, this.url, true);
    this.xhr.responseType = this.responseType;
    this.xhr.timeout = this.timeout;
    this.applyHeaders();
    this.xhr.onreadystatechange = this.handleReadyStateChange.bind(
      this,
      resolve
    );
    this.xhr.addEventListener("timeout", event =>
      resolve(
        Err(
          err_with_matcher({
            xhr: this.xhr,
            type: RequestErrType.Timeout,
            event,
          })
        )
      )
    );
    this.xhr.addEventListener("abort", event =>
      resolve(
        Err(
          err_with_matcher({
            xhr: this.xhr,
            type: RequestErrType.Abort,
            event,
          })
        )
      )
    );
    this.xhr.addEventListener("error", event =>
      resolve(
        Err(
          err_with_matcher({
            xhr: this.xhr,
            type: RequestErrType.XhrErr,
            event,
          })
        )
      )
    );
    for (let l of this.xhr_evt_listeners) {
      this.xhr.addEventListener(l.type, l.listener, l.options);
    }
    for (let l of this.upload_evt_listeners) {
      this.xhr.upload.addEventListener(l.type, l.listener, l.options);
    }
    this.xhr.send(this.data);
  }

  public request(): Promise<Result<Response<T>, ClientError<T>>> {
    if (this.req_sent) {
      return this.promise;
    }
    this.promise = new Promise<Result<Response<T>, ClientError<T>>>(
      this.execute.bind(this)
    );
    this.req_sent = true;
    return this.promise;
  }

  public static create<T = any>(
    method: string,
    url: string,
    options: RequestOptions = {}
  ): Client<T> {
    const client = new Client<T>(method, url);
    if (options.query) {
      client.addQueryObj(options.query);
    }
    if (options.data) {
      client.setData(options.data);
    }
    if (options.responseType) {
      client.responseType = options.responseType;
    }
    if (options.headers) {
      client.setHeaders(options.headers);
    }
    if (options.timeout) {
      client.timeout = options.timeout;
    }
    return client;
  }

  public static get<T = any>(
    url: string,
    query?: QStr.QueryObject
  ): Promise<Result<Response<T>, ClientError<T>>> {
    return Client.create<T>("GET", url, { query }).request();
  }

  public static post<T = any>(
    url: string,
    data: RequestData,
    options: RequestOptions = {}
  ): Promise<Result<Response<T>, ClientError<T>>> {
    options.data = data;
    return Client.create<T>("POST", url, options).request();
  }

  public static put<T = any>(
    url: string,
    data: RequestData,
    options: RequestOptions = {}
  ): Promise<Result<Response<T>, ClientError<T>>> {
    options.data = data;
    return Client.create<T>("PUT", url, options).request();
  }

  public static patch<T = any>(
    url: string,
    data: RequestData,
    options: RequestOptions = {}
  ): Promise<Result<Response<T>, ClientError<T>>> {
    options.data = data;
    return Client.create<T>("PATCH", url, options).request();
  }

  public static delete<T = any>(
    url: string,
    query?: QStr.QueryObject
  ): Promise<Result<Response<T>, ClientError<T>>> {
    return Client.create<T>("DELETE", url, { query }).request();
  }

  public static setResponseType(type: XMLHttpRequestResponseType): void {
    Client.ResponseType = type;
  }
}
