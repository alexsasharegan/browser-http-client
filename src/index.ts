import { Client } from "./xhr/client"
import { Status } from "./xhr/status"

type Dictionary<T> = { [key: string]: T }

const compatChecks: Dictionary<boolean> = {
  XMLHttpRequest: "XMLHttpRequest" in window,
  Promise: "Promise" in window,
  FormData: "FormData" in window,
  "Array.isArray": typeof Array.isArray == "function",
  "Object.assign": typeof Object.assign == "function",
  "Object.keys": typeof Object.keys == "function"
}

for (const key in compatChecks) {
  if (!compatChecks.hasOwnProperty(key) || compatChecks[key]) {
    continue
  }
  const err = new Error(`[browser-http-client] Required ${key} is not available in the browser`)
  if (typeof window.onerror == "function") {
    throw err
  } else {
    console.error(err)
  }
}

// Export sub-modules for ESM
export { Client, Status }

// Export a wrapper object for browser.
const Http = { Client, Status }
export default Http
