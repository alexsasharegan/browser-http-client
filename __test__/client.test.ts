import { Client } from "../src/index";
import { RequestErrType } from "../src/xhr/client";

const base = `https://jsonplaceholder.typicode.com`;
const mock = {
  url: suffix => base + suffix,
  wait: ms => new Promise(resolve => setTimeout(resolve, ms)),
};
Client.Timeout = 2000;

interface Geo {
  lat: string;
  lng: string;
}

interface Address {
  street: string;
  suite: string;
  city: string;
  zipcode: string;
  geo: Geo;
}

interface Company {
  name: string;
  catchPhrase: string;
  bs: string;
}

interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
}

describe("Client", async () => {
  it("should exist", async () => {
    expect(Client).toBeDefined();
  });

  it("should make a GET request", async () => {
    let result = await Client.get<User>(mock.url("/users/1"));
    expect(result.unwrap().data).toMatchSnapshot();
    expect(result.unwrap().data.id).toBeDefined();
    expect(typeof result.unwrap().data.id == "number").toBe(true);
    expect(typeof result.unwrap().data.name == "string").toBe(true);
    expect(
      Object.entries(result.unwrap().headers).map(([k, v]) => [k, typeof v])
    ).toMatchSnapshot();
  });

  it("should make POST request", async () => {
    let data = {
      title: "foo",
      body: "bar",
      userId: 1,
    };
    let result = await Client.post(mock.url("/posts"), data);
    expect(result.is_ok()).toBe(true);
    expect(result.unwrap().data).toMatchSnapshot();
  });

  it("should make PUT request", async () => {
    let data = {
      id: 1,
      title: "foo",
      body: "baz",
      userId: 1,
    };
    let result = await Client.put(mock.url("/posts/1"), data);
    expect(result.is_ok()).toBe(true);
    expect(result.unwrap().data).toMatchSnapshot();
  });

  it("should make PATCH request", async () => {
    let data = {
      title: "bar",
    };
    let result = await Client.patch(mock.url("/posts/1"), data);
    expect(result.is_ok()).toBe(true);
    expect(result.unwrap().data).toMatchSnapshot();
  });

  it("should make DELETE request", async () => {
    let result = await Client.delete(mock.url("/posts/1"));
    expect(result.is_ok()).toBe(true);
    expect(result.unwrap().data).toMatchSnapshot();
  });

  it("should abort with cancel token", async () => {
    let request = Client.create("GET", mock.url("/posts"));
    let cancel = request.getCancelToken();

    enum Code {
      Abort = 1,
      HttpStatusErr = 2,
      Timeout = 3,
      XhrErr = 4,
    }

    // Spawn off the request
    let p1 = request.send();
    let p2 = request.send();
    // Cancel it right away
    cancel();
    // Await the promise that was canceled
    let result = await p1;

    expect(result.is_err()).toBe(true);
    // extract the wrapped error value
    let err = result.unwrap_err();
    expect(err.type).toBe(RequestErrType.Abort);
    expect(
      err.match({
        Abort: () => Code.Abort,
        HttpStatusErr: () => Code.HttpStatusErr,
        Timeout: () => Code.Timeout,
        XhrErr: () => Code.XhrErr,
      })
    ).toBe(Code.Abort);

    // Check the promises are the same
    expect(p1).toBe(p2);
  });
});
