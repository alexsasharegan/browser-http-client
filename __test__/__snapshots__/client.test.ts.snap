// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client should make DELETE request 1`] = `Object {}`;

exports[`Client should make PATCH request 1`] = `
Object {
  "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
  "id": 1,
  "title": "bar",
  "userId": 1,
}
`;

exports[`Client should make POST request 1`] = `
Object {
  "body": "bar",
  "id": 101,
  "title": "foo",
  "userId": 1,
}
`;

exports[`Client should make PUT request 1`] = `
Object {
  "body": "baz",
  "id": 1,
  "title": "foo",
  "userId": 1,
}
`;

exports[`Client should make a GET request 1`] = `
Object {
  "address": Object {
    "city": "Gwenborough",
    "geo": Object {
      "lat": "-37.3159",
      "lng": "81.1496",
    },
    "street": "Kulas Light",
    "suite": "Apt. 556",
    "zipcode": "92998-3874",
  },
  "company": Object {
    "bs": "harness real-time e-markets",
    "catchPhrase": "Multi-layered client-server neural-net",
    "name": "Romaguera-Crona",
  },
  "email": "Sincere@april.biz",
  "id": 1,
  "name": "Leanne Graham",
  "phone": "1-770-736-8031 x56442",
  "username": "Bret",
  "website": "hildegard.org",
}
`;

exports[`Client should make a GET request 2`] = `
Array [
  Array [
    "content-type",
    "string",
  ],
  Array [
    "cache-control",
    "string",
  ],
  Array [
    "pragma",
    "string",
  ],
  Array [
    "expires",
    "string",
  ],
]
`;
